lamindb/history/_trigger_installer.py

```diff
@@ -0,0 +1,564 @@
+            for source_col, target_col in zip(
+                foreign_key_constraint.source_columns,
+                foreign_key_constraint.target_columns,
+            )
+        )
+
+        variable_name = self._create_unique_variable_name(
+            prefix="_fkey_uid_d" if is_delete else "_fkey_uid"
+        )
+
+        self._validate_table_name(foreign_key_constraint.target_table)
+
+        self.declare_variable(
+            variable_name,
+            "jsonb",
+            f"""
+coalesce(
+    (
+        SELECT {self._build_jsonb_object({c: c for c in table_uid.uid_columns})}
+        FROM {foreign_key_constraint.target_table}
+        WHERE {where_clause}
+    ),
+    {self._build_jsonb_object(dict.fromkeys(table_uid.uid_columns, "NULL"))}
+)
+""",  # noqa: S608
+        )
+
+        return variable_name
+
+    def _validate_table_name(self, table_name: str):
+        if not self.VALID_TABLE_NAME_REGEX.match(table_name):
+            raise ValueError(
+                f"Table name '{table_name}' doesn't look like a valid PostgreSQL table name"
+            )
+
+    def build(self) -> str:
+        self._validate_table_name(self.table)
+
+        self.declare_variable(
+            name="history_triggers_locked",
+            type="bool",
+            declaration="SELECT EXISTS(SELECT locked FROM lamindb_historylock LIMIT 1) AND "
+            "(SELECT locked FROM lamindb_historylock LIMIT 1)",
+        )
+
+        self.declare_variable(
+            name="latest_migration_id",
+            type="smallint",
+            declaration="SELECT MAX(id) FROM lamindb_historymigrationstate",
+        )
+
+        table_id_var = self.add_table_id_variable(table=self.table)
+        self.declare_variable(name="record_data", type="jsonb")
+        self.declare_variable(name="event_type", type="int2")
+        self.declare_variable(name="record_uid", type="jsonb")
+
+        # Build the function body so that any variables created as a side-effect of constructing
+        # record data get populated.
+
+        function_body = f"""
+    IF NOT history_triggers_locked THEN
+        IF (TG_OP = 'DELETE') THEN
+            record_data := NULL;
+            record_uid := {self._build_record_uid(is_delete=True)};
+            event_type := {HistoryEventTypes.DELETE.value};
+        ELSE
+            record_data := {self._build_record_data()};
+            record_uid := {self._build_record_uid(is_delete=False)};
+
+            IF (TG_OP = 'INSERT') THEN
+                event_type := {HistoryEventTypes.INSERT.value};
+            ELSIF (TG_OP = 'UPDATE') THEN
+                event_type := {HistoryEventTypes.UPDATE.value};
+            END IF;
+        END IF;
+
+        INSERT INTO lamindb_history
+            (id, history_migration_state_id, table_id, record_uid, record_data, event_type, created_at)
+        VALUES
+            (gen_random_uuid(), latest_migration_id, {table_id_var}, record_uid, record_data, event_type, now());
```

Comment:
Can we replace this with an 18-char base62 random string instead of a UUID? -- Also see the comment in the schema definition.

```sql
CREATE OR REPLACE FUNCTION base62(n_char integer) RETURNS text AS $$
DECLARE
    alphabet text := '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
    result text := '';
    i integer;
BEGIN
    FOR i IN 1..n_char LOOP
        result := result || substr(alphabet, 1 + floor(random() * 62)::integer, 1);
    END LOOP;
    RETURN result;
END;
$$ LANGUAGE plpgsql;
```